src/main/java/com/regnosys/rosetta/common/compile/JavaCSourceCancellableCompiler.java

```diff
@@ -0,0 +1,182 @@
+    public static final int DEFAULT_THREAD_POLL_INTERVAL_MS = 100;
+    public static final int DEFAULT_MAX_COMPILE_TIMEOUT_MS = 300_000;
+
+    private final int threadPollIntervalMs;
+    private final int maxCompileTimeoutMs;
+    private final ExecutorService executorService;
+    private final boolean useSystemClassPath;
+    private final boolean deleteOnError;
+    private final boolean isVerbose;
+    private final JavaCompileReleaseFlag releaseFlag;
+    private final Path[] additionalClassPaths;
+
+    public JavaCSourceCancellableCompiler(int threadPollIntervalMs,
+                                          int maxCompileTimeoutMs,
+                                          ExecutorService executorService,
+                                          boolean useSystemClassPath,
+                                          boolean deleteOnError,
+                                          boolean isVerbose,
+                                          JavaCompileReleaseFlag releaseFlag,
+                                          Path... additionalClassPaths) {
+        this.threadPollIntervalMs = threadPollIntervalMs;
+        this.maxCompileTimeoutMs = maxCompileTimeoutMs;
+        this.executorService = executorService;
+        this.useSystemClassPath = useSystemClassPath;
+        this.isVerbose = isVerbose;
+        this.deleteOnError = deleteOnError;
+        this.releaseFlag = releaseFlag;
+        this.additionalClassPaths = additionalClassPaths;
+    }
+
+    public JavaCSourceCancellableCompiler(ExecutorService executorService,
+                                          boolean useSystemClassPath,
+                                          boolean deleteOnError,
+                                          boolean isVerbose,
+                                          JavaCompileReleaseFlag releaseFlag,
+                                          Path... additionalClassPaths) {
+        this(DEFAULT_THREAD_POLL_INTERVAL_MS,
+                DEFAULT_MAX_COMPILE_TIMEOUT_MS,
+                executorService,
+                useSystemClassPath,
+                deleteOnError,
+                isVerbose,
+                releaseFlag,
+                additionalClassPaths);
+    }
+
+    @Override
+    public JavaCompilationResult compile(List<Path> sourceJavaPaths,
+                                         Path targetPath,
+                                         Supplier<Boolean> isCancelled) throws ExecutionException, InterruptedException, TimeoutException {
+        DiagnosticCollector<JavaFileObject> diagnosticCollector = new DiagnosticCollector<>();
+        JavaCompiler.CompilationTask compilationTask = getCompilationTask(sourceJavaPaths, targetPath, diagnosticCollector);
+
+        Future<Boolean> submittedTask = executorService.submit(compilationTask);
+
+        Optional<Boolean> result = submitAndWait(submittedTask, isCancelled);
```

Comment:
Return enum from submitAndWait (try state bools are hard to under stand).

Reply:
done